쥬스 메이커 [STEP 2] sookoong, bard

bar-d opened the pull request:
안녕하세요 제이크! @jryoun1 
STEP2 PR 보냅니다~~

## 고민한 점🤔
- ### `화면전환`
    - 처음 `네비게이션`을 사용해서 화면전환을 할지, `모달` 방식을 이용해서 화면전환을 할지 고민했었습니다. 
	- `네비게이션`은 주로 정보의 깊이와 흐름을 가지는 구조에서 사용되고, `모달`은 화면 전체를 정보의 흐름에서 벗어나 전환해주는 것이라고 알고있습니다.
	- 하지만, 재고를 변경하는 것은 정보의 흐름에서 벗어나지 않고 메인 스토리보드의 정보를 흐름에 따라 변경한다 생각해 `네비게이션`을 사용하였습니다.


- ### `얼럿 타입`
	- 재고가 부족할 때 나타나는 아니오 얼럿구문은 `.cancel`와 `.destructive`중 무엇을 사용해야 할지 고민했습니다.
	- `.destructive`는 데이터를 파괴하거나 변경할 때 사용한다고 나와있는데, 아니오를 누를 경우, 데이터의 변환없이 화면을 다시 띄어주면 되는 것이어서 `.cancel`스타일을 선택하였습니다.



- ### `IBAction button`
	- 처음에는 각 버튼에 `IBAction` 이벤트를 연결해주어 총 일곱개의 `IBAction` 이벤트가 생성되었습니다.
	- 너무 많은 이벤트가 생성이 되어 비효율적이라고 생각하였습니다.
	- 이를 한곳으로 모을 수 있을까라는 생각에 `sender.tag`와 `sender.currentTitle`이라는 메소드를 알게되었고, `currentTitle`을 사용하여 적용해보았습니다.





## 질문하고 싶은 것🙋

- ### `텍스트 레이블 언래핑`
	- 텍스트 레이블을 강제 언래핑을 하는 이유는 개발자가 제대로 된 레이블에 연결을 하기 때문에 강제 언래핑을 사용하는 것으로 알고 있습니다.
	- 하지만, 프로젝트 제약사항에 `!`는 최대한 사용하지 말라는 문구가 있어서 `optional binding(?)`를 사용해 언래핑을 하였습니다.
	- UILabel이 다른 코드들을 보았을 때 강제 언래핑을 많이 사용하는데 현 프로젝트에서도 강제 언래핑을 사용해도 될까요?



- ### `error handling`
	- STEP1에서 `error handling`을 할 때, 재고가 부족하면 `localized Error`타입을 통해 에러 메세지를 출력해주었는데, STEP2를 진행하면서 출력해줄 필요가 없다고 생각하여 `FruitStore`내 `handle`함수에서 `String`으로 반환값을 주었는데, `handle`함수의 결과가 사용되지 않는다고 노란색 에러가 뜹니다.
	- 이럴 때에는 반환값을 `String`으로 주는 것이 나을까요? 아니면 출력을 하도록 그냥 두는것이 나을까요?🤔

- ### `함수 은닉화`
	- `JuiceMaker`구조체 내에 있는 `make`함수를 `ViewController`로 빼내서 사용을 하였는데, 이를 위해 `JuiceMaker`의 메소드들의 은닉화를 풀어줘야만 했습니다.
	- 은닉화를 풀지않고 하는 방법이 있을까요?

- ### `singleton`
	- 싱글톤 디자인 패턴을 적용해보려고 했는데, 안의 인스턴스 멤버를 사용할 수 없다고 합니다.
	- 만약 싱글톤 디자인 패턴을 적용시키려면 전체적인 리팩토링이 필요할까요?
	- 아니면 저희가 적용을 제대로 못시킨 걸까요?

>```swift
>extension JuiceMaker {
>    static let warehouse = JuiceMaker.fruitStore.fruitWarehouse
>}
><오류메시지>
>Cannot use instance member 'fruitStore' within property initializer; property initializers run >before 'self' is available
>```


- ### `notification`
	- 최근에 배운 `notification`을 적용해보려고 했는데 현재 상황에서는 `notification`이 사용될 필요가 없다고 생각하여 사용하지 않았습니다. 
	- 만약 `notification`을 사용한다면 `FruitStockViewController` 뷰 컨트롤러에서 데이터를 발송하면 `main ViewController`가 받는 방법으로 사용해야 될 것 같습니다.
	- 제이크의 생각은 어떠한가요?🤔

Jeon-Minsu committed:
feat: 재고수정 화면으로 Navigation 기능 추가

bar-d committed:
style:  재고수정 화면 오토레이아웃 설정

Jeon-Minsu committed:
feat: 스토리보드의 과일 수량을 나타내는 UILabel 생성 및 연결

bar-d committed:
feat: 각 쥬스를 주문하는 버튼 생성 및 연결

bar-d committed:
feat: 각 쥬스를 주문하는 버튼 연결

Jeon-Minsu committed:
refactor: JuiceMaker 모듈 내 위치한 make() 함수를 뷰컨트롤러 내로 위치 변경

bar-d committed:
feat: 각 과일의 재고 레이블을 쥬스메이커의 인스턴스 내의 과일재고 개수로 동기화하는 작업추가

bar-d committed:
feat: 각 주스를 만드는 버튼 생성
- Juice 열거형 내 Int 타입 설정
- 주스를 만드는 버튼들을 태그를 사용하여 한 버튼으로 연결

Jeon-Minsu committed:
feat: 쥬스 재료의 재고가 있는 경우 얼럿 표시 기능 구현

bar-d committed:
feat: 쥬스 재료의 재고가 없는 경우 얼럿 표시 기능 구현

Jeon-Minsu committed:
refactor: 함수 내 리팩토링 과정 수행
- FruitStore 열거형 내 handle() 함수를 문자열로 반환
- Juice 열거형 반환 타입을 Int에서 String으로 수정
- 옵셔널 강제 추출 방식(!)을 옵셔널 바인딩(?) 방식으로 변경
- 상수명을 someJuiceMaker에서 juiceMaker로 변경
- ViewController 내 은닉화 과정 수행
- if let 구문을 guard let 구문으로 통일
- ViewController 내 함수 순서 변경

bar-d committed:
docs: README.md 파일 수정

jryoun1 commented on JuiceMaker/JuiceMaker/Controller/ViewController.swift:
```diff
@@ -8,11 +8,76 @@ import UIKit
 
 class ViewController: UIViewController {
```
ViewController라는 이름은 명확하지 않은 것 같습니다!

Jeon-Minsu replied:
과일 주문에 관한 뷰컨트롤러이기 때문에 `JuiceOrderViewController`로 네이밍을 수정하였습니다.

jryoun1 commented on JuiceMaker/JuiceMaker.xcodeproj/project.pbxproj:
```diff
@@ -138,6 +138,14 @@
 		C73DAF2B255D0CDC00020D38 /* Project object */ = {
 			isa = PBXProject;
 			attributes = {
+				KnownAssetTags = (
```
tag를 사용하셨군요! 멋있습니다 👍🏻

jryoun1 commented on JuiceMaker/JuiceMaker.xcodeproj/project.pbxproj:
```diff
@@ -138,6 +138,14 @@
 		C73DAF2B255D0CDC00020D38 /* Project object */ = {
 			isa = PBXProject;
 			attributes = {
+				KnownAssetTags = (
```
질문이 있습니다🙋 
만약 프로젝트의 규모가 커지고, 협업을 하는 사람들이 늘어난다면 
1. tag를 사용하는 방식과 
2. IBAction 메소드 혹은 storyboard에서 action을 연결하는 방법 

중에 어떤 것이 더 유지보수에서 편리할 것이라고 생각하는지, 왜 그렇게 생각하는지 수꿍과 바드의 의견이 궁금합니다.

bar-d replied:
아무래도 IBAction함수를 각각 선언해주면 코드만 보아도 이해를 할 수 있지만, tag나 currentTitle을 사용한다면, 다른 파일 내에서도 어떻게 tag나 currentTitle를 달았는지 확인을 해봐야 할 것 같습니다.
IBAction함수는 각각 선언해주기에 명확하지만, 코드가 길어지고, 여러개의 버튼에 하나의 IBAction이 사용이 된다면 동일한 동작이 아닌 이상, 사용이 어려울 것 같습니다.
대신 tag나 currentTitle은 사용하는 기준이 명확해 보이지 않을수는 있지만, 하나의 IBAction에 모아서 각각의 기능으로 분기할 수 있다는 장점이 있습니다.  
이로 인해 stroryboard에서 action을 연결하는 방법이 더 유지보수에 용이할 것 같습니다

jryoun1 replied:
좋습니다👏
이 부분은 **코드로 화면을 만드는 것**과 **스토리보드로 화면을 만드는 것**도 연관이 있을 것 같네요.
물론 회사 by 회사이지만 협업을 할 때는 스토리보드를 사용하면 직접 화면을 확인해야하는 문제나, 각 view component들 사이의 constraint를 파악하는 것이 코드보다 어렵다는 단점들이 있긴합니다. 

앞으로의 프로젝트에서 스토리보드를 통해서 협업을 해보는 경험과 또 코드로 화면을 구현하는 협업을 경험을 해보시면서 직접 느껴보시는 것을 추천드립니다! 물론 둘 중에 뭐가 더 좋다 라는 정답은 당연히 없고, 둘 가지 방법 모두 사용되는 방법이니 두 방법을 잘 사용할 줄 알아야합니다 🙏🏻

jryoun1 commented on JuiceMaker/JuiceMaker/Controller/ViewController.swift:
```diff
@@ -8,11 +8,76 @@ import UIKit
+    
     override func viewDidLoad() {
         super.viewDidLoad()
         // Do any additional setup after loading the view.
```
필요하지 않는 코드같아 보이네요.

Jeon-Minsu replied:
삭제 완료했습니다.

jryoun1 commented on JuiceMaker/JuiceMaker/Controller/ViewController.swift:
```diff
@@ -8,11 +8,76 @@ import UIKit
     
+    private let juiceMaker = JuiceMaker()
+    
+    @IBOutlet weak var strawberryLabel: UILabel?
```
옵셔널의 언래핑 방법들에 대한 공부를 해보시는 걸 추천드립니다. 
그 중에서도 **암시적 언래핑(implicitly unwrapping)** 에 대해서 공부해보시면 첫 번째 질문에 대한 답변이 될 것 같네요 🙏🏻

bar-d replied:
현 제약사항에 (!)를 최대한 지양하라고 나와있는데, 현재 같은 경우는 개발자가 무조건 제대로 된 값을 연결한다는 가정하에 암시적 언래핑을 하는 것으로 알고있습니다
공부를 해보고 질문을 드렸는데, 프로젝트 제약사항 때문에 여쭤본거였습니다.

jryoun1 replied:
아하! 제약사항때문에 여쭤보신 거였군요.
최대한 지양하라고 제약이 있지만 만약 써야하는데 명확한 이유가 있다면 사용하고 그 근거를 말씀해주시면 될 것 같습니다🙏🏻 

IBOutlet에서 만큼은 오히려 강제 언래핑을 권장한다고 볼 수도 있을 것 같네요. 
괜찮은 관점인 것 같아서 한 번 보시는 것도 괜찮을 것 같아서 링크 남깁니다! [참고](https://jeong9216.tistory.com/191)

jryoun1 commented on JuiceMaker/JuiceMaker/Controller/ViewController.swift:
```diff
@@ -8,11 +8,76 @@ import UIKit
+    private let juiceMaker = JuiceMaker()
+    
+    @IBOutlet weak var strawberryLabel: UILabel?
+    @IBOutlet weak var bananaLabel: UILabel?
```
이 label들은 외부에서 접근이 가능한데 외부에서 접근 가능해야하는 이유가 있을까요? 👀

bar-d replied:
외부에서 접근할 일이 없기에 `private` 접근제한자를 걸어주었습니다.

jryoun1 commented on JuiceMaker/JuiceMaker/Controller/ViewController.swift:
```diff
@@ -8,11 +8,76 @@ import UIKit
+    }
     
-}
+    @IBAction func makeJuiceButton(_ sender: UIButton) {
```
이 함수도 외부에서 접근이 가능해야하는 이유가 있나요? 👀

bar-d replied:
위와 동일하게 `private` 접근제한자를 걸어주었습니다.🙃

jryoun1 commented on JuiceMaker/JuiceMaker/Controller/ViewController.swift:
```diff
@@ -8,11 +8,76 @@ import UIKit
+        updateFruitStock()
     }
     
+    private func updateFruitStock() {
```
접근제어를 잘 사용해주셨네요 👏

jryoun1 commented on JuiceMaker/JuiceMaker/Controller/ViewController.swift:
```diff
@@ -8,11 +8,76 @@ import UIKit
     
-}
+    @IBAction func makeJuiceButton(_ sender: UIButton) {
+        guard let juice = Juice(rawValue: sender.currentTitle ?? "") else { return }
```
sender의 currentTitle 프로퍼티를 사용해보셨군요👏

jryoun1 commented on JuiceMaker/JuiceMaker/Controller/ViewController.swift:
```diff
@@ -8,11 +8,76 @@ import UIKit
+    }
     
-}
+    @IBAction func makeJuiceButton(_ sender: UIButton) {
```
함수명이 `makeJuiceButton`이면 쥬스버튼을 만든다인데 이 함수의 역할은 쥬스버튼을 만들어주는 함수인가요? 👀

Jeon-Minsu replied:
해당 함수의 역할은 쥬스를 제조하는 함수입니다. UIButton 타입을 sender로 받기에 Button을 추가하였는데, `makeJuice`로 명명하는 것이 더 명확해 보입니다.

jryoun1 commented on JuiceMaker/JuiceMaker/Controller/ViewController.swift:
```diff
@@ -8,11 +8,76 @@ import UIKit
+        updateFruitStock()
+    }
+    
+    private func make(_ juice: Juice) {
```
함수명이 아주 명확하니 좋네요 👍🏻

jryoun1 commented on JuiceMaker/JuiceMaker/Controller/ViewController.swift:
```diff
@@ -8,11 +8,76 @@ import UIKit
+            try juiceMaker.checkPossibilityOfMaking(juice)
+            showSuccessAlert(juice)
+        } catch (let error) {
+            juiceMaker.fruitStore.handle(error)
```
이 부분이 2번째 질문하신 부분이군요!
```swift
enum FruitStoreError: Error {
    case missingProduct
    case outOfStock
    case invalidAmount
}

extension FruitStoreError: LocalizedError {
    public var errorDescription: String? {
        switch self {
        case .missingProduct:
            return NSLocalizedString("해당 상품이 없습니다", comment: "Description of missing Product")
        case .outOfStock:
            return NSLocalizedString("재고가 부족합니다", comment: "Description of out of stock")
        case .invalidAmount:
            return NSLocalizedString("잘못된 수량입니다", comment: "Description of invalid amount")
        }
    }
}
```
말씀해주신 것처럼 handle()은 위에서 case에 맞는 string을 반환하고 있습니다. 
이때 출력을 한다면 해당 출력문은 사용자에게 보일까요 아니면 수꿍과 바드에게(xcode의 console에서) 보일까요? 
**누구에게 보여주고 싶은 것인지를 우선적으로 정해야할 것 같습니다.**

에러와 warning은 전혀 다른 것입니다. 
말씀해주신 노란색⚠️은 Xcode의 컴파일 warning입니다. 
그렇다면 warning과 error의 차이는 무엇일까요? 👀

Jeon-Minsu replied:
`Error`는 기본적으로 컴파일 도중 발생하는 `Complie Error`, 컴파일은 되었으나 실행 도중 발생하는 `Runtime Error`, 그리고 애매하지만 논리적인 문제로 예기치 못한 결과를 나타내는 `Logic Error`로 분류하는데, 이들은 공통적으로 문제를 해결하지 않으면 원활히 실행시킬 수 없는 상태를 의미합니다.
반면에, `warning`은 프로그램 실행은 가능하지만 잠재적인 문제를 가지고 있는 상태를 의미합니다. 앞서 저희가 사용한 `handle()` 구문은 선언은 하였지만 이를 사용하지 않아 잠재적인 문제를 가지고 있음을 나타냅니다.

저희가 활용하고자 한 각각의 에러에 대한 출력값은 저희에게(`xcode`의 `console`에서)만 보입니다. 이를 사용자 입장에서도 노출되고자 한다면 `ViewController` 내 얼럿 구문을 이용하여 표시할 수 있을 것 같습니다. 그 예로, 재고 부족 시 재고 수정을 요구하는 얼럿을 생성하여 사용자에게 에러 결과를 노출시켜보았습니다.

jryoun1 replied:
두 개의 차이를 잘 정리해주셨네요 👏 
넵 잘 적용해주셨군요!

jryoun1 commented on JuiceMaker/JuiceMaker/Controller/ViewController.swift:
```diff
@@ -8,11 +8,76 @@ import UIKit
+        updateFruitStock()
+    }
+    
+    private func make(_ juice: Juice) {
```
이 부분이 3번째로 질문하신 JuiceMaker 내부의 make함수가 viewController로 나온 부분이군요. 
Juice를 만드는 역할은 누가하는 걸까요? 🤔 
ViewController로 make() 가 나온 이유가 있을까요?? 👀

`juiceMaker.make()` 함수를 호출하면 `check~` 함수들의 은닉화도 풀지 않아도 되고, 
JuiceMaker도 자신의 역할을 수행하지 않을까요?

bar-d replied:
쥬스를 만드는 역할은 쥬스메이커가 해야된다고 생각합니다🥲  
처음 함수를 구현했을 때 `JuiceMaker`내 `make()`함수에 `throw`를 해주지 않았어서 ViewController부분으로 빼내주었는데, 나중에 `throw`가 빠진 것을 확인하고 다시 함수를 생성해주었습니다.  
은닉화 작업도 다시 수행하였습니다.

jryoun1 commented on JuiceMaker/JuiceMaker/Controller/ViewController.swift:
```diff
@@ -8,11 +8,76 @@ import UIKit
+        present(failAlert, animated: true)
+    }
+    
+    @IBAction func editStock(_ sender: UIBarButtonItem) {
```
이 함수의 역할이 stock을 수정하는 역할을 수행하는건가요? 
@IBAction 함수의 네이밍을 보내는 이벤트의 종류에 따라서 지어보는 것은 어떨까요?
예를들면 네비게이션 오른쪽 버튼을 누르는 @IBAction에 대한 네이밍은 이런식이 되면 현재 어떤 버튼과 어떤 이벤트에 의해서 이 함수가 호출되는지 더 명확해 질 것 같네요.  
```swift
@IBaction func navigationRightButtonTapped(_ sender: UIBarButtonItem) {}
@IBaction func navigationRightButtonDidTouchUpInside(_ sender: UIBarButtonItem) {}
```

Jeon-Minsu replied:
재고 수정을 수행하는 버튼이 아니라 재고 수정 화면으로 넘어가는 버튼임을 뒤늦게 깨달았습니다. '재고수정버튼 클릭시'라고 네이밍을 하면 더욱 복잡해질 것 같아 위와 같이 `navigationRightButtonTapped`로 이름을 수정하는 것이 더욱 좋아보입니다.

jryoun1 commented on JuiceMaker/JuiceMaker/Controller/ViewController.swift:
```diff
@@ -8,11 +8,76 @@ import UIKit
+
+    private func showSuccessAlert(_ juice: Juice) {
+        let successAlert = UIAlertController(title: "\(juice.name) 쥬스 나왔습니다!\n 맛있게 드세요!",
+                                             message: nil,
```
UIAlertController를 선언하실 때는 UIAlertAction과는 다르게 줄바꿈을 사용하셨네요👍🏻
특별한 이유가 있을까요?

Jeon-Minsu replied:
`UIAlertController` 내부에 매개변수가 많아 가독성 향상을 위해 줄바꿈을 사용한 반면, `UIAlerAction`은 매개변수가 간단하여 줄바꿈 없이 처리하였습니다. 코드 컨벤션의 통일성을 위하여 줄바꿈 없이 모두 처리하였습니다.

jryoun1 replied:
앗 😅 꼭 통일하실 필요는 없습니다. 그냥 줄바꿈에 특별한 기준이 있는지 궁금해서 여쭤봤습니다.

개발자의 경우에는 물론 개발자 바이 개발자지만 코드가 가로가 긴 것보다는 세로로 긴 것을 더 선호한다고 합니다.
예를 들어 32인치 모니터를 사용하는 사람은 한줄에 코드를 볼 수 있지만, 13인치 맥북으로 보는 사람은 두줄, 세줄처럼 보일수도 있기 때문에 가로보다는 세로를 선호하는 것 같다고 합니다. 

물론 이러한 규칙도 협업하는 사람과 정할 수 있습니다! (팀의 코딩컨벤션이라고하죠)
예를 들어 한 줄에 120줄이 넘지 않도록 하는 규칙을 정할수도 있고, 넘을 경우 어떻게 줄바꿈을 할지 
그리고 이것을 강제하는 방법들도 있는데 나중에 시간이 있을 때 swiftLint에 대해서 한 번 보시면 좋을 것 같습니다🙏🏻

jryoun1 commented on JuiceMaker/JuiceMaker/Controller/ViewController.swift:
```diff
@@ -8,11 +8,76 @@ import UIKit
+                                             preferredStyle: .alert)
 
+        let ok = UIAlertAction(title: "확인", style: .default, handler: nil)
+
```
showFailAlert()에서는 줄바꿈이 따로 없던 것 같은데 줄바꿈에 대한 기준이 따로 있을까요? 👀

Jeon-Minsu replied:
위 코멘트와 동일합니다.

jryoun1 commented on JuiceMaker/JuiceMaker/Controller/ViewController.swift:
```diff
@@ -8,11 +8,76 @@ import UIKit
+        let ok = UIAlertAction(title: "확인", style: .default, handler: nil)
+
+        successAlert.addAction(ok)
+
```
여기도 줄바꿈이 있는데 확인부탁드려요~

Jeon-Minsu replied:
코드 컨벤션 통일성을 위하여 상수는 상수끼리, 함수는 함수끼리 모아 수정하였습니다.

jryoun1 commented on JuiceMaker/JuiceMaker/Controller/ViewController.swift:
```diff
@@ -8,11 +8,76 @@ import UIKit
+    }
+    
+    private func showFailAlert() {
+        let failAlert = UIAlertController(title: "재료가 모자라요.\n 재고를 수정할까요?",
```
UIAlertController와 UIAlertAction을 잘 사용하셨네요 👏

jryoun1 replied:
UIAlertController의 title 및 action도 선언하신 error의 종류에 따라서 다르게 나타날 수 있게 구현할 수 있지 않을까요? 
2번 질문에서 handle()의 반환 값이 사용되지 않는다고 하셔서 해당 메서드의 반환 string을 title로 사용하거나 error의 종류에 따라서 다른 alert를 띄워줄 수도 있을 것 같다는 생각이 드네요!

이 부분은 기획서에는 없는 부분이니까 참고만 하셔도 될 것 같아요~ 🙏🏻

jryoun1 commented on JuiceMaker/JuiceMaker/Controller/ViewController.swift:
```diff
@@ -8,11 +8,76 @@ import UIKit
+        let yes = UIAlertAction(title: "예", style: .default) { action in
+            guard let editMenu = self.storyboard?.instantiateViewController(withIdentifier: "FruitStockViewController") else { return }
+            
+            self.navigationController?.pushViewController(editMenu, animated: true)
```
이 부분은 editStock 함수의 내부와 동일한 코드네요! 
함수로 빼서 중복된 코드를 줄일 수 있을 것 같아보입니다.

bar-d replied:
`goToFruitStockViewController`라는 함수로 빼주었습니다~

jryoun1 commented on JuiceMaker/JuiceMaker/Model/FruitStore.swift:
```diff
@@ -6,7 +6,7 @@
 // 과일 저장소 타입
 class FruitStore {
-     private var fruitWarehouse: Dictionary<Fruit, Int> = [
+     private(set) var fruitWarehouse: Dictionary<Fruit, Int> = [
```
private(set) 을 사용하셨군요! 멋집니다 👍🏻
private과 어떤 차이가 있고 왜 변경하셨나요? 👀

bar-d replied:
`private`은 읽고 쓰기가 가능하지만, `private(set)`은 쓰지 않고 외부에서 읽기만 가능하고 내부에서만 수정이 가능하도록 하는 코드입니다.
현 상황에서는 내부에서 수정이 필요없어 보여 변경해보았습니다🙃

jryoun1 commented on JuiceMaker/JuiceMaker/Model/FruitStore.swift:
```diff
@@ -6,7 +6,7 @@
 
 // 과일 저장소 타입
```
불필요한 주석은 제거되면 좋을 것 같네요!

Jeon-Minsu replied:
삭제 완료하였습니다.

jryoun1 commented on JuiceMaker/JuiceMaker/Model/Juice.swift:
```diff
@@ -5,12 +5,33 @@
-    case mango
-    case mangoKiwi
+enum Juice: String {
+    case strawberry = "딸기쥬스 주문"
```
Juice의 종류를 모아둔 열거형 타입인 것 같은데 strawberry가 `딸기쥬스 주문`인 것이 어색한 것 같네요.
sender의 currentTitle을 사용하기 위해서 수정된 것 같은데, 만약 새로운 동료가 들어와서 코드를 본다면 명확하지 않아서 헷갈릴 것 같습니다.

tag, currentTitle을 사용하는 방법과 IBAction 함수를 사용하는 것의 장단점은 무엇이 있을까요? 👀

bar-d replied:
아무래도 `IBAction`함수를 각각 선언해주면 코드만 보아도 이해를 할 수 있지만, `tag`나 `currentTitle`을 사용한다면, 다른 파일 내에서도 어떻게 `tag`나 `currentTitle`를 달았는지 확인을 해봐야 할 것 같습니다.  
`IBAction`함수는 각각 선언해주기에 명확하지만, 코드가 길어지고, 여러개의 버튼에 하나의 `IBAction`이 사용이 된다면 동일한 동작이 아닌 이상, 사용이 어려울 것 같습니다.  
대신 `tag`나 `currentTitle`은 사용하는 기준이 명확해 보이지 않을수는 있지만, 하나의 `IBAction`에 모아서 각각의 기능으로 분기할 수 있다는 장점이 있습니다.

Jeon-Minsu replied:
명확성을 확보하기 위하여 `makeJuice` 함수 내에 `switch` 구문을 사용하여 각각의 `currentTitle`에 맞는 과일을 배정했습니다. 허나 이 과정에서 `currentTitle`에 기존의 과일 주문 `title` 외에 다양한 구문이 들어올 수 있어 `default` `Case`를 처리해주어야 했습니다. 그 과정에서 반환 타입이 `Juice` 타입이므로, `Juice` 타입에 `unknownFruit`이라는 새로운 `case`를 생성해내야했고, 연쇄적으로 `Fruit`, `Juice`, `CostOfJuiceProduction` 열거형 까지 해당 `case`를 수정해주어야 했습니다. 

일단 위의 방법으로 코드를 수정하는 과정에서 하나의 기능으로 여러 가지 버튼들을 작동시키는 것이 좋은 코드라고 생각했으나, 명확성과 유지보수 차원에서 각각의 버튼에 `IBAction`을 활용하는 것이 더욱 좋을 것 같다고 판단하게 되었습니다. 이러한 판단이 맞는지 제이크의 의견이 궁금합니다.

`tag`, `currentTitle`과 더불어, `IBAction` 함수의 장단점은 위의 코멘트를 통해 정리해보았습니다.

jryoun1 replied:
많이 이야기를 했던 것 같아서 지겨우실수도 있지만 코드에는 정답이 없습니다. 
따라서 제가 드리는 의견은 참고만 해주시고, 결정은 수꿍 바드가 해주시면 좋을 것 같네요🙇🏻‍♂️

지금은 버튼들이 하나의 기능을 하지만 기획에서 버튼들 중 하나의 역할을 다르게 요청을 한다면 아마 currentTitle의 종류에 따라서 분기처리가 해당 함수에 들어가야하고, 그렇게 된다면 하나의 함수에서 해결하는 메리트가 없어지게 될 것입니다. 
이러한 부분들도 고려하고, 수꿍과 바드가 말씀해주신 것처럼 명확성과 유지보수 차원에서도 각각의 버튼을 IBAction으로 만들어주는 방법이 저도 괜찮다고 느껴지네요.

jryoun1 commented on JuiceMaker/JuiceMaker/Model/JuiceMaker.swift:
```diff
@@ -8,20 +8,11 @@
 struct JuiceMaker {
     let fruitStore = FruitStore()
     
-    func make(_ juice: Juice) {
```
쥬스메이커의 역할은 쥬스를 만들 수 있는지 검사만 하면 되는 걸까요? 🤔

bar-d replied:
쥬스를 만들수 있는지 검사하고 쥬스메이커가 쥬스를 만들어야 하기 때문에 위 코멘트에서 말한 것처럼 다시 함수를 넣어주었습니다!

jryoun1 reviewed (changes requested):
수꿍 바드 step2도 고생하셨습니다🙏🏻

기획서에는 navigationBar의 색상이 있고, 정확한 글씨 크기에 대한 명세는 없지만 훨씬 큰 것 같네요! 

<img width="400" alt="image" src="https://user-images.githubusercontent.com/45090197/166485341-30983331-73cb-4bae-9923-71f013d892db.png">

반면에 수꿍과 바드가 구현하신 쥬스메이커는 navigationBar의 색상과 글씨크기가 작은 것 같습니다.
<img width="500" alt="image" src="https://user-images.githubusercontent.com/45090197/166485263-b9fcd980-a767-4e62-8027-55a28f4d2c68.png">


## 고민한 점
1. 정보의 흐름에서 벗어난다는 의미는 어떤 걸까요? 🤔

예를 들어 앱에서 비밀번호 찾으러 가는 것은 정보의 흐름에서 벗어나는 걸까요?
또 다른 예시로 iPhone 설정 앱에서 각각의 설정에 들어가는 것은 정보의 흐름에서 벗어나는 건가요?

2. Alert의 타입

[Apple 공식문서](https://developer.apple.com/documentation/uikit/uialertaction/style)를 보면
-  `cancel` 은 해당 액션을 취소하고 아무런 변화없이 그대로 놔두는 경우 사용
- `destructive`는 데이터를 변화시키거나 삭제하는 경우 사용

수꿍과 바드가 말씀하신 것처럼 나와있네요. 저도 cancel이 나을 것 같네요. 
주로 destructive는 내가 작성한 글을 삭제, 수정하는 것과 같이 데이터의 변경이 있는 경우 사용자에게도 경고를 심어주기 위해서 사용합니다.

추가로 질문이 있습니다. 🙋 
cancel과 destructive 스타일을 적용했을 때 무엇이 달라질까요? 👀

3. IBAction, tag, currentTitle 의 비교

이 부분에 대해서는 아래 코드리뷰에서 답변 남겼습니다. 🙏🏻

---

## 질문
> 1. 텍스트 레이블 언래핑, 2. error handling, 3. 함수 은닉화

아래 리뷰에서 답변 남겨드렸습니다.

> 4. singleton

싱글톤 디자인 패턴도 고민을 하셨군요 👍🏻
먼저 싱글톤 패턴을 적용하려는 이유가 있을까요? 
생각보다 이 부분은 지금 이해하기에는 어려울 수도 있어서 지금은 싱글톤이 어떤 것인지 느낌만 알고 앞으로 프로젝트를 해나가면서 struct, class에 대한 차이와 singleton 패턴은 어떤 객체에 사용해야 하는지를 공부하시는 것이 이해에 도움이 될 것 같습니다.

해당 오류 메세지는 어떠한 뜻일까요? 
`Cannot use instance member 'fruitStore' within property initializer; property initializers run >before 'self' is available` 

instance member인 `fruitStore`는 property initializer에서 사용할 수 없다. property initializer가 self가 가능하기 전에 실행되고 있다. 라고 의미인 것 같네요.
이를 이해하기 위해서는 타입 프로퍼티, 인스턴스 프로퍼티에 대한 먼저 공부가 필요할 것 같네요🙏🏻 
둘의 차이점과 어느 시점에 프로퍼티의 값이 초기화 되는지(메모리에 올라가는지)를 찾아보면 될 것 같네요.

이 부분도 개인적으로는 초반에 이해하기 어려울 수 있어서 나중으로 넘기는 걸 추천드립니다 🙂
모든 걸 초반에 이해하려고 하면 지칠수 있으니까 큰 개념들은 어떤것인지말 알고 조금 나중으로 미뤄도 됩니다!

> 5. notification

새롭게 배운 것을 사용해 보는 것은 아주 좋습니다. 👏  찐개발자이시군요~ 
그러나 PR을 보내거나 할 때는 해당 방법을 왜 사용했는지에 대한 이유가 있으면 좋을 것 같습니다.
viewController들끼리 통신하는 방법에는 어떠한 방법들이 있는지 알아보고 각각을 직접 구현해보거나 공부해 본 뒤, **어떠한 이유 때문에 해당 방법을 사용했다** 라는 식으로의 근거를 가지고 코드를 작성하는 습관이 나중에 큰 도움이 될 것이라고 생각합니다. 

수꿍과 바드가 말씀하신 것처럼 notification을 사용하는 것도 하나의 방법이 될 수 있겠네요 🙂

Jeon-Minsu committed:
style: 네이밍 변경, 주석 제거 및 줄바꿈 처리

bar-d committed:
refactor:  접근제한자 설정 및 함수 리팩토링
- JuiceMaker내 make 함수 재생성 및 JuiceOrderViewController 내 tryMaking함수 생성
- FruitStockViewController 뷰컨트롤로 이동하는 함수 생성

Jeon-Minsu committed:
refactor: makeJuice() 함수 내 RawValue로 접근하는 방식에서 switch 구문을 통한 접근 방식으로 수정

Jeon-Minsu commented:
안녕하세요 제이크! @jryoun1 

이번에도 좋은 코멘트를 많이 남겨주셔서 감사합니다!

먼저 언급해주신, 색상과 글씨 크기는 Step3 오토레이아웃 적용 과정을 통해 진행해보고자 합니다!

## 고민한 점
1. 정보의 흐름에서 벗어난다는 의미는 어떤 걸까요?

>앱에서 비밀번호를 찾는 과정은 정보의 흐름에서 벗어나는 것이라고 생각합니다. 로그인을 통하여 원하는 화면으로 넘어가는 것이 올바른 정보의 흐름이라고 판단하고, 이를 위해 비밀번호를 찾으러 가는 것은 자연스러운 정보의 흐름을 위한 행동은 맞지만 그 흐름을 끊고 가는 것이라고 생각합니다.
iPhone 설정 앱에 각각의 설정으로 들어가는 행동은 설정이라는 큰 틀에서 세부 설정으로 가는 자연스러운 흐름이기 때문에 정보의 흐름에 벗어나지 않았다고 생각합니다.

2. Alert의 타입

> `cancel` 버튼은 파란색 글씨에 굵은 글씨, `destructive`는 빨간색 글씨로 변하나 굵은 글씨로는 변하지 않습니다.
`destructive` 버튼 같은 경우 액션을 하였을 때, 콘텐츠를 삭제한다거나 파괴적인 일을 하므로 버튼이 잘보이게 빨간 굵은 글씨로 표현이 됩니다. `destructive` 버튼이 있다면 사용자가 안전하게 나올수 있도록 `cancel`버튼을 제공하여야 합니다.
추가로, `HIG`에서는 '예','아니오'와 같은 표현을 사용하지 말라고 하였지만, 요구사항에 '예,' '아니오'로 표현하라해서 이에 따라 코드를 작성했습니다.

3. IBAction, tag, currentTitle 의 비교

>코멘트에 의견을 남겨보았습니다.

## 질문

4-1. Singleton

> 싱글톤 패턴은 공통된 기능을 다양한 곳에서 사용하고자 할 때 고정된 메모리 영역을 얻으면서 객체의 인스턴스를 오직 1개만 생성하는 방법으로 알고 있습니다. `JuiceMaker` 내의 `fruitStore` 인스턴스 내 프로퍼티인 `fruitWarehouse` 호출과정을 간단히 하기 위하여 하나의 공동 인스턴스를 호출하는게 어떨까라는 생각에서 시작하여 위의 과정을 진행해보았습니다. 하지만, `JuiceMaker`는 `struct`로 `singleton` 디자인 패턴을 적용하기에 모순되었다는 점을 깨달았고, 또한 타입 내 인스턴스 프로퍼티를 호출할 수 없다는 것을 깨달았습니다.

4-2. 타입 프로퍼티, 인스턴스 프로퍼티

> 인스턴스 프로퍼티는 구조체 혹은 클래스 내에서 갖는 일반적인 상수 혹은 변수를 의미하고, 타입 프로퍼티는 구조체 혹은 클래스 내에서 `static`(`override` 불가능), `class`(`override` 가능) 등과 같은 용어를 사용하여 상수 혹은 변수를 나타내는데, 이를 호출하기 위해서는 인스턴스에서 접근하는 것이 아니라 타입에서만 접근이 가능합니다. 
인스턴스 프로퍼티는 새로운 인스턴스가 생성될 때마다 새로운 프로퍼티가 같이 생성되지만, 타입 프로퍼티는 해당 타입에 속한 단 하나의 프로퍼티만 생성합니다.
인스턴스 프로퍼티는 인스턴스가 생성되면 메모리에 올라가는 방식이지만, 타입 프로퍼티는 인스턴스 생성과 전혀 상관 없이 타입 프로퍼티를 최초 호출하면 메모리에 올라가 초기화가 이루어져 그 이후에는 어디서든 해당 프로퍼티를 공유합니다.

5. Notification

> 데이터 전달의 방법 중 하나로 `Notification`을 학습하여 이를 활용해보고자 하는 의욕이 컸습니다. 처음 저희는 `ViewController` 내에서 `label`, `button`이 서로 떨어져 있는 점에 주목하여 이 둘 사이에 정보를 전달하기 위해서는 `NotificationCenter`를 통해 이벤트를 감지하고 알림을 송수신하여야 된다고 생각했습니다. 하지만 동일한 `ViewController` 내에 있는 경우 내부 함수의 구현만으로도 충분히 서로 다른 버튼과 레이블 간 데이터 전달이 가능하고 `Notification`을 사용했을 때보다 더욱 간결하고 명확한 코드를 작성할 수 있다고 판단하였습니다. 또한 `Notification`을 사용했을 때는 코드가 많아지면 추적이 어려워지고 이는 유지보수를 어렵게 하여 `Notification`을 사용하지 않고 함수 내부 표현을 통해 기능을 수행하였습니다. 


## 추가 질문

1. 서로 다른 VC간 Notification 메모리 문제

Step3 기능이 게시가 되어 PR을 기다리는 과정에서 간단하게 Step3 기능을 구현해보고자 노력했습니다. 그러던 중, 두번째 뷰컨트롤러에 `Observer`를 달고 첫번째 뷰컨에서 `push`를 하면서 `post`를 하게 되면 데이터 전달이 되지 않는다는 사실을 알게 되었습니다. 문제점을 찾아보니 Observer가 메모리 상에 없기 때문임을 알게 되었습니다. 만약 Step3에서 `notification`을 이용하고자 한다면 앞선 문제를 해결하기 위해서 `Observer`를 첫번째 뷰컨트롤러에서도 인식할 수 있게 하기 위하여 네비게이션을 통한 `push` 방식이 아닌 모달을 통한 `Present` 방식으로 정보를 전달해야 하는 것인지 궁금합니다.

jryoun1 commented:
@Jeon-Minsu @bar-d 
수꿍 바드 고생하셨습니다~ PR리뷰를 확인하시고, 수정 및 코멘트를 다 작성하셨다면 
아래 사진의 re-review를 요청하는 버튼을 눌러주시면 명확하게 re-review를 요청하신건지 알 수 있습니다🙇🏻‍♂️ 
<img width="1306" alt="image" src="https://user-images.githubusercontent.com/45090197/166867698-d142156a-8ef2-4ae9-933a-894120664007.png">

코멘트 주신 부분에 추가 코멘트를 남겨두었습니다~ step3에서 반영해도 될 것 같아서 이번 step2은 이만 머지하겠습니다 🙏🏻
즐거운 어린이날 보내세요~ 👦

----

> 먼저 언급해주신, 색상과 글씨 크기는 Step3 오토레이아웃 적용 과정을 통해 진행해보고자 합니다!

넵 좋습니다👍🏻

## 고민한점
1. 너무 잘 알고 계시는 군요 ㅎㅎ 👍🏻
2. 이미 H.I.G도 확인해보셨군요 👏 Apple에서는 이렇게 글씨색을 통해서도 사용자들에게 통일된 UX를 제공하려고 하고, 이를 개발자들에게는 .destructive, .cancel 을 통해서 구현할 수 있게 했습니다. 앞으로도 iOS 사용자들에게 친숙한 UX를 제공하기 위해서 H.I.G를 자주 보시면서 익숙해지는 걸 추천드립니다 🙇🏻‍♂️
3. 확인했습니다 👍🏻

## 질문
질문에 대한 답변들을 근거와 함께 잘 달아주셨군요 ㅎㅎ 감사합니다 🙇🏻‍♂️

## 추가질문
글로 작성해주신 것만으로는 정확히 어떤 상황인지는 잘 이해가 되지 않는군요 😭
notification의 경우에는 viewcontroller의 modal, push의 여부와 상관없이 

1. post하기 전에 observer가 register 되고
2. notification의 identifier가 동일하다면 

사용에 문제가 없습니다.
말씀해주신 부분은 아마 위에서 1번을 만족하지 않기 때문에 문제가 발생하는 것 같네요.
2번째 vc가 observer를 register하기 전에 1번째 vc에서 post를 하고 있기 때문에 메모리에 없다고 나온 것 같네요.
